File: CS/WindowsApplication3/DateTapeEdit/RepositoryItemDateTapeEdit.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using System.ComponentModel;
using System.Drawing;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using System.Reflection;

namespace WindowsApplication3
{
    [UserRepositoryItem("RegisterCustomEdit")]
    public class RepositoryItemDateTapeEdit : RepositoryItem
    {
        static RepositoryItemDateTapeEdit() { RegisterCustomEdit(); }

        public RepositoryItemDateTapeEdit() { }

        public const string CustomEditName = "DateTapeEdit";

        public override string EditorTypeName { get { return CustomEditName; } }

        public static void RegisterCustomEdit()
        {

            Image img = null;
            try
            {
                img = (Bitmap)Bitmap.FromStream(Assembly.GetExecutingAssembly().
                  GetManifestResourceStream("DevExpress.CustomEditors.CustomEdit.bmp"));
            }
            catch
            {
            }
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(CustomEditName,
              typeof(DateTapeEdit), typeof(RepositoryItemDateTapeEdit),
              typeof(DateTapeViewInfo), new DateTapePainter(), true, img));
            
        }

        public string GetCurrentDayValue(DateInfoType datePart, DateTime dt)
        {
            if (datePart == DateInfoType.Day)
                return dt.ToString("M");
            if (datePart == DateInfoType.Hour)
                return dt.Hour.ToString();
            return dt.Minute.ToString();
        }

        public string GetPreviousDayValue(DateInfoType datePart, DateTime dt)
        {
            dt = GetPreviousDate(datePart, dt);
            if (datePart == DateInfoType.Day)
                return dt.ToString("M");
            if (datePart == DateInfoType.Hour)
                return dt.Hour.ToString();
            return dt.Minute.ToString();
        }

        public string GetNextDayValue(DateInfoType datePart, DateTime dt)
        {
            dt = GetNextDate(datePart, dt);
            if (datePart == DateInfoType.Day)
                return dt.ToString("M");
            if (datePart == DateInfoType.Hour)
                return dt.Hour.ToString();
            return dt.Minute.ToString();
        }

        public DateTime GetPreviousDate(DateInfoType datePart, DateTime dt)
        {
            return GetDate(datePart, dt, -1);
        }

        public DateTime GetNextDate(DateInfoType datePart, DateTime dt)
        {
           return GetDate(datePart, dt, 1);
        }

        public DateTime GetDate(DateInfoType datePart, DateTime dt, int val)
        {
            try
            {
                if (datePart == DateInfoType.Day)
                    return dt.AddDays(val);
                if (datePart == DateInfoType.Hour)
                    return dt.AddHours(val);
                return dt.AddMinutes(val);
            }
            catch { }    
            return dt;
        }

        public string GetDayValueByPosition(DateInfoType datePart, PositionType positionType, DateTime dt)
        {
            if (positionType == PositionType.Top)
                return GetPreviousDayValue(datePart, dt);                       
            if (positionType == PositionType.Middle)
                return GetCurrentDayValue(datePart, dt);
            return GetNextDayValue(datePart, dt);
        }

    }
}

File: VB/WindowsApplication3/DateTapeEdit/RepositoryItemDateTapeEdit.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows.Forms
Imports System.ComponentModel
Imports System.Drawing
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports System.Reflection

Namespace WindowsApplication3
	<UserRepositoryItem("RegisterCustomEdit")> _
	Public Class RepositoryItemDateTapeEdit
		Inherits RepositoryItem
		Shared Sub New()
			RegisterCustomEdit()
		End Sub

		Public Sub New()
		End Sub

		Public Const CustomEditName As String = "DateTapeEdit"

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return CustomEditName
			End Get
		End Property

		Public Shared Sub RegisterCustomEdit()

			Dim img As Image = Nothing
			Try
				img = CType(Bitmap.FromStream(System.Reflection.Assembly.GetExecutingAssembly(). GetManifestResourceStream("DevExpress.CustomEditors.CustomEdit.bmp")), Bitmap)
			Catch
			End Try
			EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(CustomEditName, GetType(DateTapeEdit), GetType(RepositoryItemDateTapeEdit), GetType(DateTapeViewInfo), New DateTapePainter(), True, img))

		End Sub

		Public Function GetCurrentDayValue(ByVal datePart As DateInfoType, ByVal dt As DateTime) As String
			If datePart = DateInfoType.Day Then
				Return dt.ToString("M")
			End If
			If datePart = DateInfoType.Hour Then
				Return dt.Hour.ToString()
			End If
			Return dt.Minute.ToString()
		End Function

		Public Function GetPreviousDayValue(ByVal datePart As DateInfoType, ByVal dt As DateTime) As String
			dt = GetPreviousDate(datePart, dt)
			If datePart = DateInfoType.Day Then
				Return dt.ToString("M")
			End If
			If datePart = DateInfoType.Hour Then
				Return dt.Hour.ToString()
			End If
			Return dt.Minute.ToString()
		End Function

		Public Function GetNextDayValue(ByVal datePart As DateInfoType, ByVal dt As DateTime) As String
			dt = GetNextDate(datePart, dt)
			If datePart = DateInfoType.Day Then
				Return dt.ToString("M")
			End If
			If datePart = DateInfoType.Hour Then
				Return dt.Hour.ToString()
			End If
			Return dt.Minute.ToString()
		End Function

		Public Function GetPreviousDate(ByVal datePart As DateInfoType, ByVal dt As DateTime) As DateTime
			Return GetDate(datePart, dt, -1)
		End Function

		Public Function GetNextDate(ByVal datePart As DateInfoType, ByVal dt As DateTime) As DateTime
		   Return GetDate(datePart, dt, 1)
		End Function

		Public Function GetDate(ByVal datePart As DateInfoType, ByVal dt As DateTime, ByVal val As Integer) As DateTime
			Try
				If datePart = DateInfoType.Day Then
					Return dt.AddDays(val)
				End If
				If datePart = DateInfoType.Hour Then
					Return dt.AddHours(val)
				End If
				Return dt.AddMinutes(val)
			Catch
			End Try
			Return dt
		End Function

		Public Function GetDayValueByPosition(ByVal datePart As DateInfoType, ByVal positionType As PositionType, ByVal dt As DateTime) As String
			If positionType = PositionType.Top Then
				Return GetPreviousDayValue(datePart, dt)
			End If
			If positionType = PositionType.Middle Then
				Return GetCurrentDayValue(datePart, dt)
			End If
			Return GetNextDayValue(datePart, dt)
		End Function

	End Class
End Namespace
